File: src/rac/services/skill.py
"""Skill installation — `rac skill install` (v0.10.5).

:func:`install_skills` is the reusable capability behind the CLI adapter: it
owns resource loading, the never-overwrite refusal, parent-directory creation,
and the result model. With no name it installs every bundled skill
all-or-nothing; with a name it installs exactly that one. Skills land at the
documented Claude Code project discovery path,
``<dir>/.claude/skills/<name>/SKILL.md``.

Failure contract:

- unknown skill name      → :class:`~rac.core.skills.SkillNotFound` (usage
  error, exit 2 — mirrors the templates convention)
- existing skill file(s)  → :class:`SkillFileExists` (refused before any write;
  every existing file is left untouched — exit 1)
- missing packaged skill  → :class:`~rac.core.skills.SkillResourceMissing`
  (operational; a broken installation)

A nonexistent target directory is a usage error the CLI rejects before this
service runs (exit 2), matching every other path-taking command.
"""

from __future__ import annotations

from dataclasses import dataclass
from pathlib import Path

from rac.core.skills import SkillNotFound, available_skills, load_skill
from rac.errors import RACError


class SkillFileExists(RACError):
    """One or more target skill files already exist; RAC never overwrites.

    A no-name install collects every existing target and reports them all
    before anything is written, so the refusal is all-or-nothing.
    """

    def __init__(self, paths: list[str]):
        # ``.paths`` is part of the contract: callers (and tests) read the full
        # refused set, and the message differs for one path vs. many.
        self.paths = paths
        if len(paths) == 1:
            message = f"{paths[0]} already exists; rac skill install never overwrites"
        else:
            listing = "\n".join(f"  - {p}" for p in paths)
            message = (
                f"{len(paths)} skill files already exist; "
                f"rac skill install never overwrites:\n{listing}"
            )
        super().__init__(message)


@dataclass
class InstalledSkill:
    """One installed skill (stable JSON contract, ADR-007)."""

    skill: str
    path: str
    bytes_written: int

    def to_dict(self) -> dict:
        # bytes_written is tracked for callers but deliberately absent from the
        # JSON shape (pinned by test_installation_json_contract).
        return {"skill": self.skill, "path": self.path}


@dataclass
class SkillInstallation:
    """Result of a `rac skill install` run (stable JSON contract, ADR-007)."""

    skills: list[InstalledSkill]

    def to_dict(self) -> dict:
        return {
            "schema_version": "1",
            "installed": True,
            "skills": [skill.to_dict() for skill in self.skills],
        }


def install_skills(target_dir: str, skill_name: str | None = None) -> SkillInstallation:
    """Write bundled skills into ``target_dir``'s Claude Code skill path.

    With ``skill_name`` ``None`` every bundled skill is installed
    all-or-nothing; with a name, exactly that skill. Parent directories are
    created as needed and an existing skill file is never overwritten.

    Raises :class:`~rac.core.skills.SkillNotFound` for an unregistered name,
    :class:`SkillFileExists` when any target already exists, and
    :class:`~rac.core.skills.SkillResourceMissing` when a packaged resource is
    absent.
    """
    if skill_name is not None and skill_name not in available_skills():
        raise SkillNotFound(skill_name)
    names = available_skills() if skill_name is None else [skill_name]

    skills_root = Path(target_dir) / ".claude" / "skills"
    destinations = {name: skills_root / name / "SKILL.md" for name in names}

    # Load every resource and check every destination before writing anything:
    # a refusal (existing file) or a broken install (missing resource) must
    # never leave a partial installation behind. The mapping is in registry
    # order, which is the order the refused-path list is reported in.
    contents = {name: load_skill(name) for name in names}
    existing = [str(dest) for dest in destinations.values() if dest.exists()]
    if existing:
        raise SkillFileExists(existing)

    installed: list[InstalledSkill] = []
    for name, dest in destinations.items():
        content = contents[name]
        dest.parent.mkdir(parents=True, exist_ok=True)
        dest.write_bytes(content)
        installed.append(InstalledSkill(skill=name, path=str(dest), bytes_written=len(content)))
    return SkillInstallation(skills=installed)

File: rust/rac-engine/src/skill.rs
//! Bundled agent skills — `decided skill` (PORT-CONTRACT.d/15).
//!
//! Port of `src/asdecided/core/skills.py` (registry + resource loading) and
//! `src/asdecided/services/skill.py` (`install_skills`). The packaged `SKILL.md`
//! resources are embedded verbatim from `rust/decided-engine/assets/skills/`,
//! vendored byte-identical copies of the Python package files — a unit test
//! below pins that identity, because the installed file must be
//! byte-identical to what the oracle installs (skill brief, landmine 1).
//!
//! `SkillResourceMissing` (a broken Python installation) has no Rust
//! equivalent: embedded resources cannot be absent from a linked binary.

use std::path::Path;

use crate::walk::py_join;

/// One bundled skill: name and one-line description, registry order.
pub struct SkillSpec {
    pub name: &'static str,
    pub description: &'static str,
}

/// Bundled skills, in registry order (`BUNDLED_SKILLS`). `install` with no
/// name installs all of them; `list` enumerates them.
pub const BUNDLED_SKILLS: [SkillSpec; 4] = [
    SkillSpec {
        name: "decided-artifacts",
        description: "Author and maintain AsDecided Markdown artifacts with the decided CLI.",
    },
    SkillSpec {
        name: "decided-review",
        description: "Review an AsDecided corpus and work findings worst-first.",
    },
    SkillSpec {
        name: "decided-import",
        description: "Reformat one document into one valid AsDecided artifact, with human review.",
    },
    SkillSpec {
        name: "decided-capture",
        description: "Capture a new decision or requirement into a valid AsDecided artifact.",
    },
];

/// The embedded `SKILL.md` bytes, index-aligned with [`BUNDLED_SKILLS`].
pub(crate) const SKILL_BYTES: [&[u8]; 4] = [
    include_bytes!("../assets/skills/decided-artifacts/SKILL.md"),
    include_bytes!("../assets/skills/decided-review/SKILL.md"),
    include_bytes!("../assets/skills/decided-import/SKILL.md"),
    include_bytes!("../assets/skills/decided-capture/SKILL.md"),
];

/// `available_skills()` — bundled skill names, registry order.
pub fn available_skills() -> Vec<&'static str> {
    BUNDLED_SKILLS.iter().map(|s| s.name).collect()
}

fn skill_bytes(name: &str) -> Option<&'static [u8]> {
    BUNDLED_SKILLS
        .iter()
        .position(|s| s.name == name)
        .map(|i| SKILL_BYTES[i])
}

/// One installed skill (`InstalledSkill`; `bytes_written` is in the oracle's
/// model but deliberately absent from its JSON, so it is not carried here).
pub struct InstalledSkill {
    pub skill: String,
    pub path: String,
}

/// Result of a `decided skill install` run.
pub struct SkillInstallation {
    pub skills: Vec<InstalledSkill>,
}

/// The failure contract of `install_skills`, message-shaped like the oracle.
pub enum SkillInstallError {
    /// `SkillNotFound` — unregistered name (CLI usage error, exit 2).
    NotFound(String),
    /// `SkillFileExists` — refused before anything is written (exit 1).
    FileExists(String),
    /// Filesystem write failure (the oracle would raise `OSError`; carried
    /// so the CLI can fail loudly instead of pretending success).
    Io(String),
}

/// `install_skills(target_dir, skill_name)` — write bundled skills into
/// `<dir>/.claude/skills/<name>/SKILL.md`.
///
/// With no name every bundled skill is installed all-or-nothing: every
/// target path is checked BEFORE any write, and one collision refuses the
/// whole installation with nothing written (existing paths listed in
/// registry order). Emitted paths are `str(Path(dir) / ...)` — the caller's
/// `--dir` normalized by pathlib, never abspath'd (landmine 6).
pub fn install_skills(
    target_dir: &str,
    skill_name: Option<&str>,
) -> Result<SkillInstallation, SkillInstallError> {
    if let Some(name) = skill_name {
        if skill_bytes(name).is_none() {
            return Err(SkillInstallError::NotFound(format!(
                "unknown skill: {name} (available: {})",
                available_skills().join(", ")
            )));
        }
    }
    let names: Vec<&str> = match skill_name {
        Some(name) => vec![name],
        None => available_skills(),
    };

    // Check every destination first, then write — a refusal never leaves a
    // partial installation behind.
    let destinations: Vec<String> = names
        .iter()
        .map(|name| py_join(target_dir, &[".claude", "skills", name, "SKILL.md"]))
        .collect();
    let existing: Vec<&str> = destinations
        .iter()
        .filter(|dest| Path::new(dest.as_str()).exists())
        .map(String::as_str)
        .collect();
    if !existing.is_empty() {
        let message = if existing.len() == 1 {
            format!("{} already exists; decided skill install never overwrites", existing[0])
        } else {
            let listing: Vec<String> = existing.iter().map(|p| format!("  - {p}")).collect();
            format!(
                "{} skill files already exist; decided skill install never overwrites:\n{}",
                existing.len(),
                listing.join("\n")
            )
        };
        return Err(SkillInstallError::FileExists(message));
    }

    let mut installed: Vec<InstalledSkill> = Vec::new();
    for (name, dest) in names.iter().zip(&destinations) {
        let content = skill_bytes(name).expect("registered skill");
        let path = Path::new(dest.as_str());
        if let Some(parent) = path.parent() {
            std::fs::create_dir_all(parent)
                .map_err(|e| SkillInstallError::Io(format!("{e}: {}", parent.display())))?;
        }
        std::fs::write(path, content)
            .map_err(|e| SkillInstallError::Io(format!("{e}: {dest}")))?;
        installed.push(InstalledSkill {
            skill: (*name).to_string(),
            path: dest.clone(),
        });
    }
    Ok(SkillInstallation { skills: installed })
}

#[cfg(test)]
mod tests {
    use super::*;

    /// The embedded resources must be byte-identical to the Python package
    /// files the retirement oracle installs.
    #[test]
    fn embedded_bytes_equal_python_package_files() {
        for (i, spec) in BUNDLED_SKILLS.iter().enumerate() {
            let py_path = format!(
                "{}/../../src/asdecided/skills/{}/SKILL.md",
                env!("CARGO_MANIFEST_DIR"),
                spec.name
            );
            let py_bytes = std::fs::read(&py_path)
                .unwrap_or_else(|e| panic!("cannot read {py_path}: {e}"));
            assert_eq!(
                py_bytes, SKILL_BYTES[i],
                "embedded {} differs from the Python package file",
                spec.name
            );
        }
    }

    #[test]
    fn registry_order_and_names() {
        assert_eq!(
            available_skills(),
            vec!["decided-artifacts", "decided-review", "decided-import", "decided-capture"]
        );
    }
}
